lib/PhasarLLVM/Pointer/LLVMPointsToSet.cpp

```diff
@@ -201,64 +221,70 @@ void LLVMPointsToSet::computeFunctionsPointsToSet(llvm::Function *F) {
     return;
   }
   // check if we already analyzed the function
-  if (AnalyzedFunctions.find(F) != AnalyzedFunctions.end()) {
+  if (auto [Unused, Inserted] = AnalyzedFunctions.insert(F);
+      !Inserted || F->isDeclaration()) {
     return;
   }
   LOG_IF_ENABLE(BOOST_LOG_SEV(lg::get(), DEBUG)
                 << "Analyzing function: " << F->getName().str());
-  AnalyzedFunctions.insert(F);
 
   llvm::AAResults &AA = *PTA.getAAResults(F);
   bool EvalAAMD = true;
 
   // taken from llvm/Analysis/AliasAnalysisEvaluator.cpp
   const llvm::DataLayout &DL = F->getParent()->getDataLayout();
 
-  llvm::SetVector<llvm::Value *> Pointers;
-  llvm::SmallSetVector<llvm::CallBase *, 16> Calls;
-  llvm::SetVector<llvm::Value *> Loads;
-  llvm::SetVector<llvm::Value *> Stores;
-
-  for (auto &I : F->args()) {
-    if (I.getType()->isPointerTy()) { // Add all pointer arguments.
-      Pointers.insert(&I);
-    }
-  }
-
-  for (llvm::inst_iterator I = inst_begin(*F), E = inst_end(*F); I != E; ++I) {
-    if (I->getType()->isPointerTy()) { // Add all pointer instructions.
-      Pointers.insert(&*I);
+  auto getSizeOf = [&DL](const llvm::Value *V) {
+    if (!V->getType()->isPointerTy()) {
+      std::cerr << "ERROR: " << llvmIRToString(V) << " is no pointer!\n";
     }
-    if (EvalAAMD && llvm::isa<llvm::LoadInst>(&*I)) {
-      Loads.insert(&*I);
+    llvm::Type *I1ElTy = V->getType()->getPointerElementType();
+    return I1ElTy->isSized() ? DL.getTypeStoreSize(I1ElTy)
+                             : llvm::MemoryLocation::UnknownSize;
+  };
+  auto mayAlias = [&AA](const llvm::Value *V1, uint64_t V1Size,
+                        const llvm::Value *V2, uint64_t V2Size) {
+    return AA.alias(V1, V1Size, V2, V2Size) != llvm::AliasResult::NoAlias;
```

Comment:
Is there anyway to parallelize the alias calls?  AA.alias doesn't look thread safe.

Reply:
As far as I know, AA does not guarantee thread safety (the accesser methods to AAResults are not even `const` and "could" potentially change the state of the result) and LLVM for the most part is also not thread safe. From a brief glimps into the implementation I could not find anything that "should" not work, however, everything there and in phasar builds on LLVMContext and not ThreadSafeContext, so I guess we don't have any guarantees.

One thing we could try to explore this a bit further, is to build phasar (and probably LLVM) with clangs TSA and just try to run it in parallel, see what breaks. However, in the end, we probably need a thorough check of the code to not run into spurious behavior and threading bugs :rofl:.

So, before checking this out further, do you have profiles of an analysis that show this to be the major bottleneck? From my experience, I often found other spots in phasar that could be optimized with much less risk/work needed.